cpp/include/raft/allocator.hpp

```diff
@@ -0,0 +1,118 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuda_runtime.h>
+#include "cudart_utils.h"
+
+namespace raft {
+
+/**
+ * @brief Interface for an asynchronous device/host allocator.
+ *
+ * An implementation of this interface can make the following assumptions:
+ * - It does not need to be but it can allow async allocate and deallocate.
+ */
+class Allocator {
+ public:
+  /**
+   * @brief Asynchronously allocates a memory region.
+   *
+   * An implementation of this need to return a allocation of n bytes properly
+   * align bytes on the configured device. The allocation can optionally be
+   * asynchronous in the sense that it is only save to use after all work
+   * submitted to the passed in stream prior to the call to allocate has
+   * completed. If the allocation is used before, e.g. in another stream the
+   * behaviour may be undefined.
+   * @todo: Add alignment requirments.
+   *
+   * @param[in] n         number of bytes to allocate
+   * @param[in] stream    stream to issue the possible asynchronous allocation in
+   */
+  virtual void* allocate(std::size_t n, cudaStream_t stream) = 0;
+
+  /**
+   * @brief Asynchronously deallocates device memory
+   *
+   * An implementation of this need to ensure that the allocation that the
+   * passed in pointer points to remains usable until all work sheduled in
+   * stream prior to the call to deallocate has completed.
+   *
+   * @param[inout] p      pointer to the buffer to deallocte
+   * @param[in] n         size of the buffer to deallocte in bytes
+   * @param[in] stream    stream in which the allocation might be still in use
+   */
+  virtual void deallocate(void* p, std::size_t n, cudaStream_t stream) = 0;
+
+  virtual ~Allocator() {}
+};  // class Allocator
+
+/**
+ * @brief An explicit interface for an asynchronous device allocator.
+ *
+ * This is mostly done in order to reduce work needed in cuML codebase.
+ * An implementation of this interface can make the following assumptions,
+ * further to the ones listed in `Allocator`:
+ * - Allocations may be always on the device that was specified on construction.
+ */
+class deviceAllocator : public Allocator {
+};  // class deviceAllocator
+
+/**
+ * @brief An explicit interface for an asynchronous host allocations.
+ *
+ * This is mostly done in order to reduce work needed in cuML codebase.
+ * An implementation of this interface can make the following assumptions,
+ * further to the ones listed in `Allocator`:
+ * - Allocations don't need to be zero copy accessible form a device.
+ */
+class hostAllocator : public Allocator {
+};  // class hostAllocator
+
+/** Default cudaMalloc/cudaFree based device allocator */
+class defaultDeviceAllocator : public deviceAllocator {
+ public:
+  void* allocate(std::size_t n, cudaStream_t stream) override {
+    void* ptr = 0;
```

Comment:
Should RMM be the default device allocator? RMM could still default to cudaMalloc down the road afaik.

Reply:
The default is setup to be `cudaMalloc/cudaFree` as it comes with the least amount of dependencies. We could certainly add an `rmmDeviceAllocator` *header* in case our users would like to use rmm as their default and also adding that dependency.

Reply:
On a second thought, it does make sense to rely on RMM as our default. Updated the code.